### ISSUE TITLE
The documentation on the use of properties files and how to indicate whether task results are correct can be improved

### ISSUE THREAD

**Issue by shonfeder** [sosy-lab/benchexec#820]
Hello!

I haven't been able to figure out how to get my results included in the "Correct only" results in the generated tables. This is despite having spent quite some time with the documentation and trying to reason this out from the examples. I suspect I am missing a basic concept.,

The results are all giving a status of `true` (but the category shows up as `missing`, is that relevant? I haven't been able to figure out how to set set categories), 

![image](https://user-images.githubusercontent.com/19820422/158930566-9391c847-4ac1-4e93-b673-9c89d6323f4a.png)

but they will only appear on the table if I select "Results: All":

![image](https://user-images.githubusercontent.com/19820422/158930676-b5912974-c635-44ae-83ba-0ac2b04f39a4.png)

I thought this might have something to do with the "properties" files. I have found the documentation on them a bit hard to parse out, but I've tried to configure something like the following:

In the `.xml` config file:

```xml
  <tasks name="SetAdd">
    <include>010encoding-SetAdd-001-OOPSLA19-SetAdd.yml</include>
    <propertyFile expectedverdict="true">spec.prp</propertyFile>
  </tasks>
```

And then in the "task definition" `.yml` file:

```json
{
  "required_files" : [
    "array-encoding/*"
  ],
  "input_files" : [
    "array-encoding/SetAdd.tla"
  ],
  "properties" : [
    {
      "property_file" : "spec.prp",
      "expected_verdict" : true
    }
  ],
  "format_version" : "2.0"
}
```

Along with an empty file `spec.prp` (I can't figure out what should go in here). But this doesn't seem to have the desired effect.

Am I on the right track? Am I missing a part of the documentation that illustrates this? Or is there perhaps a clear example you could point me towards?

Once I can figure this out, I'd be very happy to contribute to some documentation that might help similarly clueless people like myself in future :)

**Comment by PhilippWendler** [comment]
> The results are all giving a status of `true` (but the category shows up as `missing`, is that relevant?

Yes, this is exactly the crucial point. If the category is `missing`, this means that BenchExec can not determine the correctness of the result. Whether the status is `true` is not important, after all, also `false` can be the correct result for a given task. The documentation for this is [here](https://github.com/sosy-lab/benchexec/blob/main/doc/run-results.md#additional-results-of-benchexec).

> I haven't been able to figure out how to set set categories),

BenchExec will determine it automatically based on the status and the expected verdict, one can not set it.

> I thought this might have something to do with the "properties" files. I have found the documentation on them a bit hard to parse out, but I've tried to configure something like the following:
> 
> In the `.xml` config file:
> 
> ```
>   <tasks name="SetAdd">
>     <include>010encoding-SetAdd-001-OOPSLA19-SetAdd.yml</include>
>     <propertyFile expectedverdict="true">spec.prp</propertyFile>
>   </tasks>
> ```

[`propertyfile` is all lowercase](https://github.com/sosy-lab/benchexec/blob/68c314cff94aeb778bd4e5285570f23579294efb/doc/benchmark.dtd#L48). I would guess that this is the actual problem and everything should work if you fix that.

Note that the `expectedverdict` attribute is not relevant here for you: It is a *filter* on the tasks that you specified, it does not define the expected verdict. I would remove it (but it is not related to your problem). The documentation for all this is in [benchmark.xml](https://github.com/sosy-lab/benchexec/blob/68c314cff94aeb778bd4e5285570f23579294efb/doc/benchmark.xml#L41-L43) and the [BenchExec documentation](https://github.com/sosy-lab/benchexec/blob/main/doc/benchexec.md#task-definition-files).

> And then in the "task definition" `.yml` file:
> 
> ```json
> {
>   "required_files" : [
>     "array-encoding/*"
>   ],
>   "input_files" : [
>     "array-encoding/SetAdd.tla"
>   ],
>   "properties" : [
>     {
>       "property_file" : "spec.prp",
>       "expected_verdict" : true
>     }
>   ],
>   "format_version" : "2.0"
> }
> ```

Looks fine in principle (just with unusual syntax).

> Along with an empty file `spec.prp` (I can't figure out what should go in here).

Doesn't matter for BenchExec. The main purpose of this file is to be given to the executed tool to tell it what to do, and we do not want to restrict what can be in it. BenchExec only uses this file to select which expected verdict from the task definition is used: The property file that is specified in the XML is given to the tool and as lookup key for expected verdict in the task definition.

Note that the both places where you specified `spec.prp` need to point to the same file, not just files with the same content. If your XML benchmark definition and your task definition happen to be in the same directory, this is the case. Otherwise you need to fix at least one relative path pointing to `spec.prp`.

> Am I on the right track? Am I missing a part of the documentation that illustrates this? Or is there perhaps a clear example you could point me towards?

Without actually trying it out, I think the only problem is the typo. I have linked some documentation above. An our [tests](https://github.com/sosy-lab/benchexec/tree/main/test/tasks) could also serve as example. You can execute them with `benchexec test/tasks/benchmark.xml`.

> Once I can figure this out, I'd be very happy to contribute to some documentation that might help similarly clueless people like myself in future :)

Great, that's always welcome!

**Comment by shonfeder** [comment]
Thank you for extremely helpful clarification, @PhilippWendler! You were of course correct about the typo being the only thing standing in my way here.

I'll have a lock at the docs now that I understand better how these three different files fit together and, and what the purpose and function of the properties file is.

**Comment by shonfeder** [comment]
I am planning to revisit this and do what I can to provide some suggestions. Sorry for the long delay!

**Comment by PhilippWendler** [comment]
I'll close this issue because I think there is no concrete action item. If you have any improvements, please just file a PR.
